### PR TITLE
Add support for publishing events using EventDataBatch

### LINKED ISSUE #4137: [Event Hubs] Should consider max link size when creating batches in EventDataCollector
Currently, the EventDataCollector does not know how large the link size is when it creates the EventData batch. The information needs to be propagated from ReactorSender.

### LINKED ISSUE #4263: [Event Hub] Add functionality that will create and send an EventDataBatch
Add an EventHubProducer method that will create EventDataBatch. Look at:

https://github.com/Azure/azure-sdk-for-js/issues/4082

### PR DESCRIPTION
**Pull request by conniey**
* Part of August Preview is to expose/support sending an `EventDataBatch`.
* Producers create `EventDataBatch` through `EventHubProducer.createBatch()` or `EventHubProducer.createBatch(BatchOptions)`.
* After populating the `EventDataBatch`, they use `EventHubProducer.send(EventDataBatch)` to push the batch to the service.

Fixes #4137 
Fixes #4263

### CODE REVIEW

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 1]
```diff
@@ -43,12 +60,25 @@
         this.eventBytes = new byte[maxMessageSize];
     }
 
+    /**
+     * Gets the number of {@link EventData events} in the batch.
+     *
+     * @return The number of {@link EventData events} in the batch.
+     */
     int getSize() {
```
Comment (jsquire) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: I worry that the the name is potentially confusing;  size could be interpreted as the size of the batch, in bytes, or the number of events in the batch.   My brain went to the first and had trouble reconciling with the code.  Maybe consider something like `numberOfEvents` or `countOfEvents`?

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 2]
```diff
@@ -22,11 +22,28 @@
...
 import java.util.Objects;
 
-/*
+/**
  * A class for aggregating EventData into a single, size-limited, batch that will be treated as a single message when
  * sent to the Azure Event Hubs service.
+ *
+ * <p><strong>Publish events using an {@link EventDataBatch}.</strong></p>
+ *
+ * Developers can create {@link EventDataBatch batches}, add the events they want in that batch, and publish these
```
Comment (jsquire) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Should we consider a way to check the current byte-size in the batch?   It's not the most accurate check due to the serialization, but we are asking developers to provide a body of raw bytes.   If I know that I only have 10 bytes left in the batch and I'm creating an event with a body size of 20, I can reason that it won't fit.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java [hunk 3]
```diff
@@ -74,16 +75,31 @@
  *
  * {@codesnippet com.azure.messaging.eventhubs.eventhubproducer.send#publisher-sendOptions}
  *
+ * <p><strong>Publish events using an {@link EventDataBatch}.</strong></p>
+ *
+ * Developers can create {@link EventDataBatch batches}, add the events they want in that batch, and publish these
+ * events together. When creating a {@link EventDataBatch batch}, developers can specify a set of {@link BatchOptions
+ * options} to configure this batch.
+ *
+ * In the scenario below, the developer is creating a networked video game. They want to receive telemetry about their
```
Comment (jsquire) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java]: I feel seen.   

*&lt;tear&gt;*

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java [hunk 4]
```diff
@@ -198,24 +256,48 @@
...
-                throw new IllegalArgumentException(String.format(Locale.US,
-                    "PartitionKey '%s' exceeds the maximum allowed length: '%s'.", partitionKey, MAX_PARTITION_KEY_LENGTH));
-            }
-        }
+        verifyPartitionKey(partitionKey);
 
-        //TODO (conniey): When we implement partial success, update the maximum number of batches or remove it completely.
         return sendLinkMono.flatMap(link -> {
-            return events.collect(new EventDataCollector(options, 1, link::getErrorContext))
+            //TODO (conniey): When we implement partial success, update the maximum number of batches or remove it completely.
```
Comment (jsquire) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java]: nit:  is this still relevant?   It's feeling more and more like we're not going to take on auto-splitting for GA.

File: sdk/eventhubs/azure-eventhubs/src/test/java/com/azure/messaging/eventhubs/BatchOptionsTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        String originalPartitionKey = "Some partition key";
+        int originalSize = 100;
+
+        String partitionKey = "A new partition key";
+        int size = 24;
+
+        BatchOptions options = new BatchOptions().partitionKey(originalPartitionKey).maximumSizeInBytes(originalSize);
+        BatchOptions clone = (BatchOptions) options.clone();
+
+        // Act
```
Comment (jsquire) [inline sdk/eventhubs/azure-eventhubs/src/test/java/com/azure/messaging/eventhubs/BatchOptionsTest.java]: *To be?  Or not to be?  THAT is the question!*

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 1 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Yeah, me too, when I first saw this class. `getNumberOfEvents` sounds like the winner.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java (hunk 3 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java]: I wrote this scenario, _just for you_. 😄

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 2 above)
Comment (ramya-rao-a) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: In the Javascript implementation, we did expose `size` as a property on the `EventDataBatch`

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 1 above)
Comment (ramya-rao-a) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Track 1 .Net SDK has this has `count`, so maybe `getCount`? 
Also, what is the use case for exposing count?

I can see the reason for size which @jsquire mentioned in https://github.com/Azure/azure-sdk-for-java/pull/4403/files#r302972436, but of what use is the count? If the purpose of an "event batch" is to ensure that we improve perf by batching multiple events under a size limit, then count shouldn't be the driving factor?

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java [hunk 6]
```diff
@@ -77,17 +78,32 @@
  *
  * {@codesnippet com.azure.messaging.eventhubs.eventhubproducer.send#publisher-sendOptions}
  *
+ * <p><strong>Publish events using an {@link EventDataBatch}.</strong></p>
+ *
+ * Developers can create {@link EventDataBatch batches}, add the events they want in that batch, and publish these
+ * events together. When creating a {@link EventDataBatch batch}, developers can specify a set of {@link BatchOptions
+ * options} to configure this batch.
+ *
+ * In the scenario below, the developer is creating a networked video game. They want to receive telemetry about their
```
Comment (ramya-rao-a) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java]: Wait, what?
We are adding sample scenarios in code?

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 1 above)
Comment (jsquire) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: We'll need at least an internal count or other way to understand if a batch is empty.   Size is not likely sufficient, as that would require understanding the envelope size (which isn't stable in .NET, it changes based on something that I don't remember).

I would ask the inverse question:  is there a reason NOT to offer a count, since it's something that batch should understand and an expected member of the intrinsic collection types of each language?

Comment (ramya-rao-a) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Definite yes to the need for an internal count.
Like the idea of using count to determine emptiness

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java (hunk 6 above)
Comment (JonathanGiles) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java]: This scenario text is in service of describing the code snippet that follows. There is no harm in this so long as it does not become a wall of text, as the focus should always be on the code sample itself.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+     * Gets the maximum size to allow for the batch of events, in bytes.
+     *
+     * @return The maximum size to allow for a single batch of events, in bytes.
+     */
+    public int maximumSizeInBytes() {
+        return maximumSizeInBytes;
+    }
+
+    /**
+     * Sets a hashing key to be provided for the batch of events, which instructs the Event Hubs service map this key to
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java]:
```suggestion
     * Sets a hashing key to be provided for the batch of events, which instructs the Event Hubs service to map this key to
```

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+     * a specific partition but allowing the service to choose an arbitrary, partition for this batch of events and any
+     * other batches using the same partition hashing key.
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java]: This paragraph is a bit hard to parse for a layperson.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+     *
+     * The selection of a partition is stable for a given partition hashing key. Should any other batches of events be
+     * sent using the same exact partition hashing key, the Event Hubs service will route them all to the same
+     * partition.
+     *
+     * This should be specified only when there is a need to group events by partition, but there is flexibility into
+     * which partition they are routed. If ensuring that a batch of events is sent only to a specific partition, it is
+     * recommended that the identifier of the position be specified directly when sending the batch.
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java]: Same with this paragraph - I can feel my eyes crossing over. It might benefit from either simplification or sample code (or a link to sample code elsewhere).

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+     *
+     * @param partitionKey The partition hashing key to associate with the event or batch of events.
+     * @return The updated {@link SendOptions} object.
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java]:
```suggestion
     * @return The updated {@link BatchOptions} object.
```

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+     * belong to the same group and should belong to the same partition.
+     *
+     * @return The partition hashing key to associate with the event or batch of events.
+     */
+    public String partitionKey() {
+        return partitionKey;
+    }
+
+    /**
+     * Creates a shallow clone of this instance. The parameters are not cloned, but they are immutable.
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/BatchOptions.java]: Don't leak implementation details in the JavaDoc
```suggestion
     * Creates a shallow clone of this instance.
```

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 2 above)
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]:
```suggestion
 * Developers can create an {@code EventDataBatch}, add the events they want into it, and publish these
```

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 12]
```diff
@@ -22,11 +22,28 @@
...
+ * events together. When creating a {@link EventDataBatch batch}, developers can specify a set of {@link BatchOptions
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: No need to link to the current class.
```suggestion
 * events together. When creating an {@code EventDataBatch}, developers can specify a set of {@link BatchOptions
```

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 1 above)
Comment (JonathanGiles) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: To be fair, this is not public API. I don't know if that is intentional or not, but given this, naming is less critical.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 13]
```diff
@@ -43,12 +60,25 @@
...
+     *
+     * @param eventData The {@link EventData} to add to the batch.
+     * @return {@code true} if the event could be added to the batch; {@code false} if the event was too large to fit in
+     *         the batch.
+     * @throws IllegalArgumentException if {@code eventData} is {@code null}.
+     * @throws AmqpException if {@code eventData} is larger than the maximum size of the {@link EventDataBatch}.
+     */
+    public boolean tryAdd(final EventData eventData) {
         if (eventData == null) {
             throw new IllegalArgumentException("eventData cannot be null");
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Use ClientLogger for throwing exceptions

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 14]
```diff
@@ -22,11 +22,28 @@
...
+ * Developers can create {@link EventDataBatch batches}, add the events they want in that batch, and publish these
+ * events together. When creating a {@link EventDataBatch batch}, developers can specify a set of {@link BatchOptions
+ * options} to configure this batch.
+ *
+ * In the scenario below, the developer is creating a networked video game. They want to receive telemetry about their
+ * users' gaming systems, but do not want to slow down the network with telemetry. So they limit the size of their
+ * {@link EventDataBatch batches} to be no larger than 256 bytes. The events within the batch also get hashed to the
+ * same partition because they all share the same {@link BatchOptions#partitionKey()}.
+ *
+ * {@codesnippet com.azure.messaging.eventhubs.eventhubproducer.send#eventdatabatch}
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: You should determine if it is better to reproduce the sample code snippet from event hub producer (as is done here), or if you should refer the user to the class 'for further code samples'.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/ReactorSender.java [hunk 15]
```diff
@@ -211,13 +204,34 @@ public String getEntityPath() {
...
+    }
+
     @Override
     public void close() {
         subscriptions.dispose();
         tokenManager.close();
         super.close();
     }
 
+
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/ReactorSender.java]: ...

File: sdk/eventhubs/azure-eventhubs/src/samples/java/com/azure/messaging/eventhubs/EventHubProducerJavaDocCodeSamples.java [hunk 16]
```diff
@@ -72,4 +73,42 @@ public void sendEventsFluxSendOptions() {
...
+     * Code snippet demonstrating how to create an {@link EventDataBatch} and send it.
+     */
+    public void sendEventDataBatch() {
+        final EventHubProducer producer = client.createProducer();
+
+        // BEGIN: com.azure.messaging.eventhubs.eventhubproducer.send#eventdatabatch
+        final Flux<EventData> telemetryEvents = Flux.just(
+            new EventData("92".getBytes(UTF_8)).addProperty("telemetry", "latency"),
+            new EventData("98".getBytes(UTF_8)).addProperty("telemetry", "cpu-temperature"),
+            new EventData("120".getBytes(UTF_8)).addProperty("telemetry", "fps")
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/samples/java/com/azure/messaging/eventhubs/EventHubProducerJavaDocCodeSamples.java]: Should we have a ctor overload for String?

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/ReactorSender.java (hunk 15 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/ReactorSender.java]: I feel compelled to add another blank line for you to find. 🤣

File: sdk/eventhubs/azure-eventhubs/src/samples/java/com/azure/messaging/eventhubs/EventHubProducerJavaDocCodeSamples.java (hunk 16 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/samples/java/com/azure/messaging/eventhubs/EventHubProducerJavaDocCodeSamples.java]: JS had an overload that took a string, but we removed that because people usually use complex data.. it was easier to create bytes this way.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 14 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Duplicated code is hard to maintain. I removed it and replaced it with a `@see.`

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 1 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: In our team channel, we've discussed.. it's going to be public now.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 17]
```diff
@@ -55,18 +55,28 @@
      *
      * @return The number of {@link EventData events} in the batch.
      */
-    int getSize() {
+    public int getCount() {
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Now that this is public API, can I register my preference for size over count? :-) It is pretty standard in Java to use 'size'.

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/models/EventHubConsumerOptions.java [hunk 18]
```diff
@@ -205,9 +204,7 @@ public int prefetchCount() {
     /**
      * Creates a shallow clone of this instance.
      *
-     * The object is cloned, but this instance's fields are not cloned. {@link Duration} and {@link String} are
-     * immutable objects and are not an issue. The implementation of {@link Retry} could be mutable. In addition, the
-     * {@link #scheduler()} set is not cloned.
+     * The object is cloned, but this instance's fields
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/models/EventHubConsumerOptions.java]: Don't leave us hanging!

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java [hunk 19]
```diff
@@ -80,7 +80,7 @@
  *
  * <p><strong>Publish events using an {@link EventDataBatch}.</strong></p>
  *
- * Developers can create {@link EventDataBatch batches}, add the events they want in that batch, and publish these
+ * Developers can create an {@link EventDataBatch}, add the events they want in into it, and publish these
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducer.java]: 'in into'

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java (hunk 17 above)
Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: Do we have getSize (number of items in batch) and GetSizeInBytes?

File: sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java [hunk 20]
```diff
@@ -22,52 +23,87 @@
...
+     *         the batch.
+     * @throws IllegalArgumentException if {@code eventData} is {@code null}.
+     * @throws AmqpException if {@code eventData} is larger than the maximum size of the {@link
+     *         EventDataBatch}.
+     */
+    public boolean tryAdd(final EventData eventData) {
         if (eventData == null) {
-            throw new IllegalArgumentException("eventData cannot be null");
+            logger.logAndThrow(new IllegalArgumentException("eventData cannot be null"));
+            return false;
```
Comment (ShivangiReja) [inline sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: If you are throwing error `"eventData cannot be null"`, then `return false` would be unreachable code ?

Comment (conniey) [reply sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataBatch.java]: The `logger.logAndThrow` method returns void, and the throw is inside that method, so the compiler doesn't know that it won't reach the next line of code.

**Comment by conniey** [comment]
@JonathanGiles Alright. I've made fixes to those late night CR errors. 😢 Hopefully it looks good now.
